### ISSUE TITLE
Figure out accessing system fonts

### ISSUE THREAD

**Issue by david-vanderson** [david-vanderson/dvui#133]
https://github.com/david-vanderson/dvui/issues/121#issuecomment-2362212259 points out that for many languages the default fonts we ship with dvui don't support enough characters.

Unsure how this should work, but it seems like we need some way to pull in system fonts at runtime, and either use them directly or as fallback when the built-in font doesn't have the characters.  This would be optional so apps that only want the fonts they choose to bundle can opt out.

**Comment by iacore** [comment]
We can load ttc/ttf at runtime. That's what nearly every graphical application does, I think.

- rasterization: most type faces are bezier paths. Most, as in not color emoji font, not svg font.
- hinting: opentype text without hinting looks awful.

I feel like using Github issue to track potential features is getting wonky. Since features depend on each other, it might be better to have the textual description of each issue saved somewhere else where we can edit collaboratively. 

Example using git repo itself to store docs: https://codeberg.org/VnPower/PixivFE/src/branch/v2/doc

At the very bare minimum, every feature need to list its dependencies, which are features that may not exist yet.

**Comment by david-vanderson** [comment]
Yes that sounds right - we can have either compiled-in fonts or load them at runtime (or both).  We'll normally have freetype2 compiled in, which will handle rasterization and hinting for us.  In some places (web), we're currently using stb_truetype for rasterization, which doesn't have hinting.

I'm also feeling a bit overwhelmed by the number of open issues we have, but I think it speaks to the growing usefulness of dvui.  Using github issues seems to be what people expect, so let's live with it a bit more and see.  If it is getting out of control we'll figure out something.

**Comment by VisenDev** [comment]
May I suggest either
- Using issue labels to categorize issues better or
- Using **Github Projects** to organize related issues and topics

There are many highly active github repos with thousands of issues - which shows that is it possible to scale the issues system as a project grows

**Comment by VisenDev** [comment]
Perhaps we could start by using some labels like 
- Planned/Accepted (Contributors plan to implement this)
- Bug
- Enhancement
- Low Priority
- High Priority

**Comment by david-vanderson** [comment]
Labels could be useful.  My preference would be for the labels to describe what action we are likely to take.  So maybe like "discussion", and "planned" (meaning enough details are sorted that we could start trying to implement it).  I think that covers all the existing issues?

**Comment by david-vanderson** [comment]
I closed #149 after finding that fonts were the main source of excess size in web-test.wasm.

The web backend should be able to load fonts at runtime like the other backends, and then we can slim down the list of compiled-in fonts.

**Comment by david-vanderson** [comment]
I think I finally have a decent plan for fonts.  It has multiple levels:

**1. Compile and run time font file loading**
This supports the usecase of any application that knows what fonts it will use.  This covers
- simple app with a few embedded fonts
- game that loads a font at runtime from a resource file

dvui already supports this, but the api can be better.  The demo has a prototype that can load a given ttf file.  Web demo loads a ttf file from hardcoded url.

**2. System font enumeration**
Support letting a user pick from fonts already on their system.  This covers
- rich text editing
- slideshows
- theme editing

We would need mutliple OS-specific backends for this functionality, but it is limited to selecting a single font.

**3. Font fallback**
Add css-like font selection, where the app/user can ask for a general kind of font with fallback options and dvui would try to load a "best fit" font it could find on the system.

**4. Glyph fallback**
If a unicode codepoint (or grapheme cluster if we get there) is not present in the specified font, then (if the app opts in) dvui would try to load a glyph from a fallback font.

**Comment by david-vanderson** [comment]
In #559 @phatchman pointed out that we'd like an easy way to get the "bold" or "italic" version of a base font.  I'm not sure where that falls - it seems like it should go in "Font fallback", but also makes sense even at level 1 where an app would preload everything.

**Comment by iacore** [comment]
Bold and italic fonts are often separate files in OpenType/TrueType.
https://www.freedesktop.org/software/fontconfig/fontconfig-user.html

A font family can contain many variants, including multiple bold and multiple italic ones.

i.e. A font family contains many files.

dvui then need to handle multiple fallback paths. Each font style + family is a path.

regular serif -> bold serif -> emoji
bold serif -> regular serif -> emoji

**Comment by iacore** [comment]
I think we should have a program that copies font files from system to ... say `zig-out` during build time ... and generate definition files based on font-config `fc-list`.

See `fc-list -f "%{family} | %{style} | %{file}\n"`

**Comment by david-vanderson** [comment]
Good info thanks!

**Comment by knightpp** [comment]
Is there a workaround to display unsupported characters? Or the only way is to wait until this issue resolved? I'd really like to display `і`, `ґ`, `ї`, but only Open Dyslexic and Pixelify can show those chars.

**Comment by RedPhoenixQ** [comment]
> Is there a workaround to display unsupported characters? Or the only way is to wait until this issue resolved? I'd really like to display `і`, `ґ`, `ї`, but only Open Dyslexic and Pixelify can show those chars.

There is no current font fallback logic outside of displaying the fallback character. You'd need to pick a font that supports all the characters.

**Comment by david-vanderson** [comment]
> Is there a workaround to display unsupported characters? Or the only way is to wait until this issue resolved? I'd really like to display `і`, `ґ`, `ї`, but only Open Dyslexic and Pixelify can show those chars.

Sorry - the only current way is like @RedPhoenixQ says - to bundle or load a font with those glyphs.  The web demo loads a Noto font dynamically, and the examples bundle the Noto font into the executable.

Let us know if you are having trouble bundling or loading (whichever you need).

**Comment by knightpp** [comment]
Oh, sorry I didn't understand how to load a font, but get it now. I tried `.Noto` predefined font, but it does not display the characters above on Linux+sdl3. 

Loading a custom Noto.ttf works!

**Comment by david-vanderson** [comment]
> Oh, sorry I didn't understand how to load a font, but get it now. I tried `.Noto` predefined font, but it does not display the characters above on Linux+sdl3.
> 
> Loading a custom Noto.ttf works!

Horay!

**Comment by david-vanderson** [comment]
Update: after #657 we have the start of font fallback (not glyph fallback), but still only a single pool of internal fonts.  We aren't yet accessing system fonts.

**Comment by iacore** [comment]
I have made a fc-list parser [here](https://git.envs.net/iacore/glyph-ranges/), to be used on Linux. Maybe it's better to include [mecha](https://github.com/Hejsil/mecha) as the parsing library here.

The original idea is from [here](https://www.1a-insec.net/blog/11-fonts-where/).

**Comment by david-vanderson** [comment]
> I have made a fc-list parser [here](https://git.envs.net/iacore/glyph-ranges/), to be used on Linux. Maybe it's better to include [mecha](https://github.com/Hejsil/mecha) as the parsing library here.
> 
> The original idea is from [here](https://www.1a-insec.net/blog/11-fonts-where/).

Thanks!

**Comment by iacore** [comment]
For references: [system default fonts](https://kagi.com/assistant/cf052905-867b-4e5d-acd0-6cc3dff4617f)

---

Default system fonts by operating system:

- **Windows 11 / Windows 10**: Segoe UI[^1]
- **macOS**: SF Pro (San Francisco Pro)[^2][^6]
- **iOS / iPadOS**: San Francisco (SF)[^3][^5]
- **Android**: Roboto[^4]
- **Linux distributions**: no single default; most desktop environments ship with open-source sans families such as Cantarell (GNOME), Noto Sans (many distros), Oxygen Sans (KDE), etc.

[^1]: [Change Default System Font in Windows 11](https://www.elevenforum.com/t/change-default-system-font-in-windows-11.8590/) (29%)
[^2]: [How do I make my default system font larg… - Apple Community](https://discussions.apple.com/thread/256088588) (19%)
[^3]: [How to Use System Fonts Ios 18? Elevate Your Design](https://wordscr.com/how-to-use-system-fonts-ios-18/) (16%)
[^4]: [Typography – Material Design 3](https://m3.material.io/styles/typography/fonts) (15%)
[^5]: [What Is Ios Default Font? Apple’s Typography Secret - WordSCR](https://wordscr.com/what-is-ios-default-font/) (11%)
[^6]: [Notes - What is the system default font? - Apple Community](https://discussions.apple.com/thread/253535537) (10%)

**Comment by david-vanderson** [comment]
> For references: [system default fonts](https://kagi.com/assistant/cf052905-867b-4e5d-acd0-6cc3dff4617f)

Good info.  I guess if we are going to use system fonts like this (querying the system UI font and using it) we'd need to figure out not just the default but how to get the name/size that the user may have changed it to?

**Comment by iacore** [comment]
dvui's current default font size is too small; smaller than 12pt (recommended by the web accessibility group). imgui's default font size is smaller than 12pt too.

What do you want the default font size to be? Do we have an embedded example? An example that shows dvui on top of another multimedia application (draw a color triangle).

I think it's better to create hard-coded themes inside this repo that let the library user decide which theme to use. Linux doesn't have "system UI", so the font choice is up to the user. Maybe there are already applications written in dvui that we can copy themes from.

**Comment by david-vanderson** [comment]
@knightpp Please don't thumbs-down something.  If there's something you don't agree with, then add a comment about why.  Thanks!

**Comment by david-vanderson** [comment]
> dvui's current default font size is too small; smaller than 12pt (recommended by the web accessibility group). imgui's default font size is smaller than 12pt too.

I think you are right.  I didn't really understand what was going on with font sizing until the recent update (to make font sizing use the em 'M' square height.  I need to look into making the default 12.

> I think it's better to create hard-coded themes inside this repo that let the library user decide which theme to use. Linux doesn't have "system UI", so the font choice is up to the user. Maybe there are already applications written in dvui that we can copy themes from.

Not sure I fully understand here.  I think most apps are either using the default dvui builtin theme (Adwaita), or they define their own, but the application developer (this is what you mean by library user right?) decides what theme dvui is using.  It would be nice to ship dvui with a larger collection of "high quality" themes, but unsure when we'll get to it.

I think most developers will be fine with that, but I suspect some would like the option for dvui to try and query the system at runtime to find the end-users' font.  (I think Linux and Windows support user-controlled UI fonts, unsure about Mac).

**Comment by knightpp** [comment]
> If there's something you don't agree with, then add a comment about why.

Oh, it's nothing. I don't have objections to the given information. Only that pasting a link feels like advertisement, especially when the mentioned service pays sanctioned countries knowingly.

**Comment by david-vanderson** [comment]
I think links are fine, and we aren't going to get into politics.  If the data at the link is relatively small, then posting the data here is helpful.

**Comment by iacore** [comment]
> Not sure I fully understand here. I think most apps are either using the default dvui builtin theme (Adwaita), or they define their own, but the application developer (this is what you mean by library user right?) decides what theme dvui is using. It would be nice to ship dvui with a larger collection of "high quality" themes, but unsure when we'll get to it.

I don't mean "high quality themes". I meant that a Linux theme is different from a MacOS theme. Different OS versions may use different font family. Essentially, we would have multiple themes to choose from. One may want to use the Windows 10 theme on Linux. We need to provide a default theme based on the target OS, but also the option to heavily customize the look of DVUI.

> sanctioned countries

I live under a different regime from yours so I have no idea what you are talking about, sorry.
